### PR TITLE
#171784086 Reset Password via email

### PR DESCRIPTION
**Pull request by Francois-MUGOROZI**
#### What does this PR do?
Enable user to reset a password when forgotten
#### Description of Task to be completed?
- Create a forgot password endpoint and send reset link to user's email
- Create reset password endpoint to change user password with new one
#### How should this be manually tested?
Give that the user has signed up (refer to signup feature); clone the repo
cd into it and run npm install then npm run dev.
using postman, send forgot password at localhost:5000/api/auth/forgotpassword with email in the body.
Refer to the API docs by browsing: http://localhost:5000/api-docs/
If email is yours, you will get a link in your inbox

#### What are the relevant pivotal tracker stories?
[171784086](https://www.pivotaltracker.com/n/projects/2438874/stories/171784086)
#### Screenshots (if appropriate)
Capture
Capture

### CODE REVIEW

**Review by benshidanny11 — DISMISSED** [review]
- Good job @Francois-MUGOROZI , 
**Feedbacks**
- I think you are importing files that don't exist in project. 
      Ex: import userService from '../services/userService'; & import sendEmail from '../helpers/email.create'; 
- In model directory of sequelize,  I think is where all models should be, example : User,...

File: server/controllers/usercontroller.js [hunk 1]
```diff
@@ -18,46 +19,98 @@ class User {
...
+    const { email } = req.body;
+    const getEmail = await userService.findByProp({ email });
+
+    if (!getEmail[0]) {
+      const error = `User with ${email} not found!`;
+      util.setError(404, error);
+      return util.send(res);
+    }
+
+    // generate token and send it to email
```
Comment (irfiacre) [inline server/controllers/usercontroller.js]: Remove the comments from the code

**Review by irfiacre — CHANGES_REQUESTED** [review]
Nice work @Francois-MUGOROZI.  Just Resolve The hound errors and make sure the tests are running.

**Review by kabundege — APPROVED** [review]
-  A tip on hound errors fix 
 ```just add a comma on the last element in a mentioned objects, it will be solved.```
 - You should squash our commits and remain with one.
 - make sure your tests are running

**Review by irfiacre — APPROVED** [review]

**Review by fordarnold — APPROVED** [review]

File: server/controllers/usercontroller.js [hunk 2]
```diff
@@ -18,48 +19,72 @@ class User {
...
 
-      util.setSuccess(200, message);
-      return util.send(res);
+    const message = 'Account was successfully verified.';
+
+    util.setSuccess(200, message);
+    return util.send(res);
+  }
+
+  static async forgotPassword(req, res) {
```
Comment (bdushimi) [inline server/controllers/usercontroller.js]: @Francois-MUGOROZI sendResetPasswordEmail is more appropriate as the function name.

File: server/controllers/usercontroller.js [hunk 3]
```diff
@@ -18,48 +19,72 @@ class User {
...
+    const { email } = req.body;
+    const resetToken = encode({ email });
+    const restUrl = `${process.env.RESET_URL}?token=${resetToken}`;
```
Comment (bdushimi) [inline server/controllers/usercontroller.js]: @Francois-MUGOROZI maybe you can change this variable name too to a name that describes what it contains

File: server/controllers/usercontroller.js [hunk 4]
```diff
@@ -18,48 +19,72 @@ class User {
...
+    const subj = 'Barefoot Nomad - Password Reset (Expires In 1 hour)';
+    const message = `Your password reset token (expires in 1 hour)<br/>Follow the link to reset it.<br/>If you didn't forget it ignore this message.<br/><br/>`;
+    const footer = `For more info or questions, contact : ${process.env.EMAIL}`;
+    const html = `<div>${message}<a href="${restUrl}" style="background-color:#028b95;color:white;padding:10px 20px;text-decoration:none">Clik Here</a><br/><br/>${footer}</div>`;
+
+    sendEmail(html, subj, email);
```
Comment (bdushimi) [inline server/controllers/usercontroller.js]: @Francois-MUGOROZI may you handle the error just in case nodemailer fails to send the email.

**Review by bdushimi — DISMISSED** [review]

**Pull request merged**